README.md
# Desafio DevOps jr PicPay

Obrigado pelo interesse em fazer parte do nosso time! Preparamos este desafio com carinho para ajudar a entender um pouco mais dos seus conhecimentos na área de DevOps/SRE

Se não entender algum conceito ou parte do problema, não é motivo para se preocupar! Queremos que faça o desafio até onde souber.

No mais, divirta-se :D

## Conteúdo do repositório
Na pasta `services` deste repositório existem 3 aplicações, um frontend que se comunica com um backend go e um em python, e estes se comunicam com um Redis para troca de informações. Tudo isso é orquestrado pelo docker-compose na raiz do repositório.

As aplicações contém falhas propositais, de código, projeto, imagem docker, etc. Embora cada aplicação funcione individualmente, o conjunto não sobe...

## O que deve ser feito?

Faça um fork deste repositório e envie uma pull request contendo:
- ajustes que fazem todas as aplicações subirem e se comunicarem
- um README contendo os seus pensamentos ao longo do projeto
- um desenho contendo os serviços que explique o funcionamento

Faça commits ao longo do processo, queremos entender o seu modo de pensar! :)

Para a entrevista, separe também anotações contendo melhorias que faria em cada aplicação e o motivo. Não envie estas anotações na pull request.

## Bibliografia recomendada
https://docs.docker.com/engine/reference/builder/

https://docs.docker.com/compose/compose-file/

https://12factor.net/

https://conventionalcommits.org/

CHANGELOG.md
## Changelog

As alterações realizadas visam corrigir a estrutura do projeto, assim como definir uma arquitetura para o sistema. 

- `web` conecta-se com `reader` e `writer` pela rede `frontend`, não havendo conexão com o armazenamento (`redis`).

- O back end, composto pelos serviços `reader` e `writer` se comunicam diretamente com o armazenamento (`redis`) para trocar informações.

### root

- Adicionado Makefile para subir e descer containers com `docker-compose`.

- Adicionado CHANGELOG.md para visualizar alterações feitas no código.

### docker-compose.yaml

- Adicionada rede `frontend`.

- Corrigido nome do serviço `redis`.

- Alterado porta para serviço `web` de 5000 para 5001.

- Corrigindo as portas dos serviços `reader` e `writer`.

- Adicionando a porta utilizada no `redis`.

- Adicionado `volumes` em todos os containers para facilitar desenvolvimento.

- Adicionando `depends_on` para seguir um fluxo dependência na subida de containers.

### services/frontend

- Corrigindo `Dockerfile`.

### services/reader

- Corrigindo `Dockerfile`.

- Adicionando arquivos `go.mod` e `go.sum` com as dependencias do projeto.

- Corrigindo endpoint `/data` onde o `client.Get` passava parametros excedentes. Era necessário passar somente uma string que seria a key: `key := client.Get("SHAREDKEY")`.

### services/writer

- Corrigindo `Dockerfile`.

- Adicionando `requirements.txt`.

reader/main.go
package main

import (
    "fmt"
    "github.com/go-redis/redis"
    "github.com/rs/cors"
    "log"
    "net/http"
)


func main() {
    redis_host := "redis"
    redis_port := "6379"
    mux := http.NewServeMux()

    mux.HandleFunc("/health", func(writer http.ResponseWriter, request *http.Request){
        if request.Method == "OPTIONS" {
            writer.WriteHeader(http.StatusOK)
            fmt.Println("GET /health (down)")
            return
        }
        fmt.Println("GET /health (up)")
        fmt.Fprintf(writer, "up")
    })

    mux.HandleFunc("/data", func(writer http.ResponseWriter, request *http.Request) {
        client := redis.NewClient(&redis.Options{Addr: redis_host+":"+redis_port})
        key := client.Get("SHAREDKEY")
        fmt.Println("GET /data (" + key.Val() + ")")
        fmt.Fprintf(writer, key.Val())
    })

    handler := cors.New(cors.Options{
        AllowedOrigins: []string{"*"},
        AllowedHeaders: []string{"*"},
    }).Handler(mux)

    log.Fatal(http.ListenAndServe(":8080", handler))

}

writer/main.py
import http.server
import socketserver
import json
import redis

class RequestHandler(http.server.SimpleHTTPRequestHandler):

    def _send_cors_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', '*')
        self.send_header('Access-Control-Allow-Headers', '*')

    def do_OPTIONS(self):
        self.send_response(200, "ok")
        self._send_cors_headers()
        self.end_headers()

    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-type", "application/json")
        self._send_cors_headers()
        self.end_headers()

        if self.path == '/health':
            self.wfile.write(bytes(f"up", "utf8"))

        return
    def do_POST(self):
        self.send_response(201)
        self.send_header("Content-type", "application/json")
        self._send_cors_headers()
        self.end_headers()
        if self.path == '/write':
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            redishost = "redis"
            redisclient = redis.Redis(host=redishost)
            redisclient.set("SHAREDKEY",post_data.decode('utf-8'))



handler_object = RequestHandler
server = socketserver.TCPServer(("", 8081), handler_object)
server.serve_forever()

frontend/js/app.js
const root = document.querySelector("#root")

const home = () => {
    const container = document.createElement('div');
    container.innerHTML = `
        <h2>Home</h2>
        `

    const readerp = container.appendChild(document.createElement('p'))
    const writerp = container.appendChild(document.createElement('p'))
    readerhealth(readerp)
    writerhealth(writerp)
    container.classList.add("mx-auto", "align-middle", "container")

    return container;
};

const reader = () => {
    const container = document.createElement('div');
    container.innerHTML = `
        <h2>Reader Screen</h2>
        `
    const child = container.appendChild(document.createElement('p'))
    readData(child)
    container.classList.add("mx-auto", "align-middle", "container")

    return container;
};

const writer = () => {
    const container = document.createElement('div');
    container.innerHTML = `
        <h2>Writer Screen</h2>
        <form id="submitdata" role="form">
          <div class="form-group">
            <label for="text">Key Content:</label>
            <input type="text" class="form-control" id="post">
          </div>
          <button type="submit" class="btn btn-default">Enviar</button>
        </form>
    `
    container.classList.add("mx-auto", "align-middle", "container")
    const formsubmit = container.querySelector("#submitdata")
    formsubmit.addEventListener('submit', writeData)
    return container;
}

const readerhealth = (readerparagraph) => {
    const url = 'http://localhost:8080/health';
    fetch(url, {
            method: "GET",
            mode: 'cors',
            headers: {
                'Access-Control-Allow-Origin': '*'
            },
        }).then(response => {
            response.text().then(body => {
                assembleStatus(readerparagraph, "Reader", body)
            }).catch(err => {
                assembleStatus(readerparagraph, "Reader", "down")
            })
        })
        .catch(error => {
            console.log(error)
            assembleStatus(readerparagraph, "Reader", "down")
        })
}

const writerhealth = (writerparagraph) => {
    const url = 'http://localhost:8081/health';
    fetch(url, {
        method: "GET",
        mode: 'cors',
        headers: {
            'Access-Control-Allow-Origin': '*'
        },
    }).then(response => {
        response.text().then(body => {
            assembleStatus(writerparagraph, "Writer", body)
        }).catch(err => {
            assembleStatus(writerparagraph, "Writer", "down")
        })
    }).catch(error => {
        assembleStatus(writerparagraph, "Writer", "down")
    })
}

const assembleStatus = (paragraph, service, status) => {
    let statusIcon = 'danger'
    let statusBadge = 'danger'
    if (status == "up") {
        statusIcon = 'success'
        statusBadge = 'success'
    }
    paragraph.innerHTML =
        service + ' service status <span class="badge badge-' + statusBadge + '">' + status + '</span>'
}

async function writeData(e) {
    e.preventDefault();

    const url = 'http://localhost:8081/write';
    fetch(url, {
        method: "POST",
        body: e.target.elements.post.value,
        mode: 'cors',
        headers: {
            'Access-Control-Allow-Origin': '*'
        },
    }).then(response => {

    }).catch(error => {
        console.log(error)
    })
}

const readData = (paragraph) => {
    const url = 'http://localhost:8080/data';
    fetch(url, {
            method: "GET",
            mode: 'cors',
            headers: {
                'Access-Control-Allow-Origin': '*'
            },
        }).then(response => {
            response.text().then(body => {
                paragraph.innerHTML = "Valor encontrado = " + body
            }).catch(err => {
                paragraph.innerHTML = `
            <div class="alert alert-warning" role="alert">
              Ocorreu um erro ao buscar a chave
            </div>
        `
            })
        })
        .catch(error => {
            console.log(error)
        })
}


const routes = {
    home: home(),
    reader: reader(),
    writer: writer(),
}

const validateHash = (hash) => hash === "" ? 'home' : hash.replace('#', '');
const init = () => window.addEventListener('hashchange', renderPage);

const renderPage = () => {
    const page = validateHash(window.location.hash);
    root.innerHTML = '';
    root.appendChild(routes[page])
}

window.addEventListener('load', () => {
    renderPage();
    init();
});